### DISCUSSION TITLE
Ai function which returns control to the User does not return response to the user

### DISCUSSION THREAD

**Discussion by dcssi** [zhudotexe/kani#19] [Q&A] [answered]
I have the following function:
@ai_function(after=ChatRole.USER)
    async def create_kani(
        self,
        name : Annotated[str, AIParam(desc="Name of the kani")],
    ):
      """Creates a new Kani."""
      if name in self.kanis:
         return "a Kani with that name already exists."
      n = Kani(self.engine)
      self.kanis[name] = n
      return f" Created Kani {name}"

When calling ai.full_round_str and the above function is executed, i get one response generated by the function formatter, but i do not get the response generated by the function it self.  The function response is added to the history ( i can get it from there), however I think it should be part of the return generated by full_round_str.

for the caller of full_round_str, where the response was generated ai  or function is irrelevant.

**Comment by zhudotexe** [comment, marked as answer]
We discussed this and will change `Kani.full_round` to emit all messages rather than only assistant messages in v0.4.0.

**Reply by dcssi** [reply]
Please also discuss the following:
In the case a function returns to the user, is the return of the function "Always" needed in the chat history ?
If not "Always" then its bit of a waste of chat tokens.
If Kani.full_round, will return to the user the function response, maybe it would be better for the user to add the response of the function to the history him self rather than kani always doing it.
I guess ai_function could have a no_history field

**Reply by zhudotexe** [reply]
Even if your function returns to the user you should give the LLM some feedback that it was successful; if you need to display different contents to the user and model you can use your own logging/print mechanism in the function body and return just a short success message to the LLM.
